doc/hare-dtm-hld.md

```diff
@@ -0,0 +1,614 @@
+
+### [r.DTM.participants]
+Hare must know or should be able to identify the participating nodes in DTM
+recovery with respect to a failed node.
+
+### [r.DTM.Spiel.interface]
+Hare invokes Motr Spiel API to send outgoing DTM messages.
+
+### [r.DTM.recover]
+Hare must be able to trigger DTM recovery for a failed node. DTM recovery
+triggers must be sent to the participating nodes only.
+
+### [r.DTM.trigger.once]
+Hare must trigger DTM recovery for a given node exactly once with respect to
+a failure. This requires all the hax process to coordinate.
+
+### [r.DTM.recovery.states]
+Hare must maintain DTM recovery states for all the nodes participating in
+DTM recovery.
+
+### [r.DTM.complete]
+Hare must wait for DTM recovery complete messages from all the participating
+nodes.
+
+### [r.notify.online]
+Hare must broadcast ONLINE ha event for the failed node on successful
+completion of DTM recovery.
+
+### [r.concurrent.failure]
+Hare must handle Hax, Consul and additional Motr process failures during
+an ongoing recovery.
+
+## Design Highlights
+Motr DTM helps to recover an incomplete meta data update transaction that was
+partially executed only on a subset of nodes due a participating process
+failure. Hare monitors and maintains the states for Motr processes and is thus
+responsible to trigger DTM recovery in case of transient failures
+i.e. Motr process restart, crash, node restart, etc. The recovery is triggered
+when the failed Motr process, mainly that hosts ioservice restarts. There are
+multiple sources through which the Hare can receive a process or node failure
+and online events, viz.: Consul and cortx-ha. It is important that Hare records
+all the state changes with respect to a Motr process from its start in order as
+they are received in order to facilitate decision making with respect to
+DTM recovery and failure handling, hare uses Consul for the same. Following
+sequence diagram illustrates a typical happy path for DTM recovery triggered
+by Hare,
+
+<!--
+@startuml
+autonumber
+systemd -> consul: start
+systemd -> hax: start
+systemd -> motr: start
+hax -> consul: populate kv
+    loop if RC
+        hax -> consul: initialize failure vector and epoch
+    end
+destroy motr
+ha -> hax: notify motr failure
+hax -> consul: update motr process state to M0_CONF_HA_PROCESS_STOPPED
+    loop if RC
+        hax -> consul: update motr process ha state to M0_NC_FAILED
+        hax -> consul: add a failvec entry for current epoch
+    end
+motr -> hax: restarts, send entrypoint request
+hax -> consul: fetch ha states for given epoch
+hax -> motr: send entrypoint reply and ha states
+hax -> hax: analyse failure vector
+hax -> consul: update motr process ha state to M0_NC_TRANSIENT
+hax -> motr: broadcast M0_NC_TRANSIENT using BQ
+motr -> hax: notify M0_CONF_HA_PROCESS_STARTING
+hax -> consul: update motr process ha state to M0_NC_RECOVERING
+hax -> motr: boradcast M0_NC_RECOVERING
+hax -> hax: wait for dtm recovery to complete
```

Comment:
>1. What if the node where such a waiting hax crash? Do we assume that m0d processes send entrypoint requests to local hax only?

Yes, m0ds connect only to their local hax process as a ha process.

> 2. What if hax process crashes and restarts?

I somehow missed that we need to handle this in LR2 version, yes we have a problem here. In Halon, when Halon satellite used to restart, Halon used to restart the corresponding motr process too to re-establish the halinks.
In LR1, because we have pacemaker dependency between hax and motr process, this case is handled. So if hax restarts, corresponding motr processes restart too.
In LR2 as well cortx-ha is using pacemaker in which we can set the hax - motr dependency. But probably it is good idea to have an independent solution in hax itself. I wonder how SAGE handles it, because they do not use pacemaker.
Created https://jts.seagate.com/browse/EOS-18514.

>As far as I understand, nobody else will poll DTM recovery status (and probably nobody will be aware that a DTM recovery has been started). Can't it cause an issue?

If you mean in context of concurrent IO, then motr handles concurrent IO during DTM recovery and Hare need not worry.
If you mean in context of concurrent Failures, then Hare will handle that as mentioned in the design via maintained HA states in Consul KV.

Reply:
Added documentation.